### DISCUSSION TITLE
Is there a special option for a polytropic equation of state?

### DISCUSSION THREAD

**Discussion by kshariff786** [PrincetonUniversity/athena#540] [Q&A]
(1) Is there a special option for a polytropic eos, where p = K rho^beta, or do I have to go the general eos route?
(2) Am I correct in understanding that the flag NON_BAROTROPIC_EOS (set in configure.py) really means "non isothermal"?

Thanks,
Karim

**Comment by c-white** [comment]
If K is a fixed constant throughout the simulation, then this is not a supported option. Actually, for beta not equal to 1, I don't know that the general eos can help; I think it assumes a two-dimensional thermodynamic space, though I'm happy to be corrected.

If K varies and you really want a gas that changes volume adiabatically (until it is shocked), then that is the adiabatic eos.

Athena only has adiabatic, isothermal, and general (non-barotropic) eos, and so NON_BAROTROPIC_EOS basically means non-isothermal. This flag is used throughout the code to indicate that the fluid has a 5D state (whether that be density / 3D velocity / pressure or density / 3D momentum / energy) rather than 4D (density / 3D velocity or density / 3D momentum), and so array sizes and loop limits are adjusted accordingly.

**Comment by msbc** [comment]
The general eos assumes a 2d thermodynamic space. You might be kludge it in
the general eos, but I wouldn't recommend it.

**Reply by kshariff786** [reply]
Thank-you for your replies Chris and Matt.  I was initially encouraged when one of the Athena++ publications in ApJ Supp. mentioned that a polytropic eos could be treated using the general eos option.  I was indeed thinking of "kludging it" in a general eos as coded below.  However, the thing I am concerned about is that an energy equation pde is being evolved when it doesn't need to be.  We want the internal energy/pressure out of the total energy pde to be consistent with the pressure from the eos.  This is far from being guaranteed.  This matters if the code uses the total energy (from the pde) to calculate the pressure (which I am sure it does for the non-isothermal option).  How difficult would it be to insert code for the polytropic case right beside code for the isothermal case (with Nhydro variables = 4)?

// ************************************************************************
// Polytropic equation of state.

Real EquationOfState::PresFromRhoEg(Real rho, Real egas) {
   return K_eos * pow(rho, beta);
}

Real EquationOfState::EgasFromRhoP(Real rho, Real pres) {
   // e internal = p / (gamma_eos - 1)
   return K_eos * pow(rho, beta) / (gamma - 1.0);
}

Real EquationOfState::AsqFromRhoP(Real rho, Real pres) {
   return gamma_eos * pres / rho;
}
// ************************************************************************

**Comment by msbc** [comment]
This should work, but you should do some tests to make sure things behave properly.

**Comment by kshariff786** [comment]
Thank-you Matt.  I can test it against Pluto.  Just one question:  Should I place this code in the code generator file?

**Reply by msbc** [reply]
You need to add a file in `src/eos/general`, e.g. `src/eos/general/polytropic.cpp`. And configure with `--eos general/polytropic`. `configure.py` might through an error here. I can't remember if Jim let me make the config script allow new EOS files (my personal branch has this feature).

**Comment by kshariff786** [comment]
Thank-you.  I'll try it!!....Karim
